app/src/main/java/supermarket/model/Offer.kt

```diff
@@ -1,3 +1,118 @@
 package supermarket.model
 
-class Offer(internal var offerType: SpecialOfferType, internal val product: Product, internal var argument: Double)
+sealed class Offer {
+    abstract fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount?
+}
+
+abstract class SingleProductOffer(val product: Product): Offer()
+
+class PercentageOffer(product: Product, val percentOff: Double) : SingleProductOffer(product) {
+    override fun getDiscount(
+        shoppingCart: ShoppingCart,
+        catalog: SupermarketCatalog
+    ): Discount? {
+        val quantityOf = shoppingCart.quantityOf(product)?.toInt() ?: return null
```

Comment:
👍 for removing this one, though I think if I were really relying on this code I'd want to come up with a better Quantity abstraction instead of just having a Double do, er, double duty. :) There are other `toInt()`s in the code that I would prefer to leave (see below).